internal/supervisor/cluster.go

```diff
@@ -0,0 +1,176 @@
+package supervisor
+
+import (
+	"github.com/mattermost/mattermost-cloud/internal/model"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/sync/semaphore"
+)
+
+// clusterStore abstracts the database operations required to query clusters.
+type clusterStore interface {
+	GetCluster(clusterID string) (*model.Cluster, error)
+	GetUnlockedClusterPendingWork() (*model.Cluster, error)
+	GetClusters(clusterFilter *model.ClusterFilter) ([]*model.Cluster, error)
+	UpdateCluster(cluster *model.Cluster) error
+	LockCluster(clusterID, lockerID string) (bool, error)
+	UnlockCluster(clusterID string, lockerID string, force bool) (bool, error)
+	DeleteCluster(clusterID string) error
+}
+
+// clusterProvisioner abstracts the provisioning operations required by the cluster supervisor.
+type clusterProvisioner interface {
+	CreateCluster(cluster *model.Cluster) error
+	UpgradeCluster(cluster *model.Cluster) error
+	DeleteCluster(cluster *model.Cluster) error
+}
+
+// ClusterSupervisor finds clusters pending work and effects the required changes.
+//
+// The degree of parallelism is controlled by a weighted semaphore, intended to be shared with
+// other clients needing to coordinate background jobs.
+type ClusterSupervisor struct {
+	store       clusterStore
+	provisioner clusterProvisioner
+	workers     *semaphore.Weighted
+	logger      log.FieldLogger
+}
+
+// NewClusterSupervisor creates a new ClusterSupervisor.
+func NewClusterSupervisor(store clusterStore, clusterProvisioner clusterProvisioner, workers *semaphore.Weighted, logger log.FieldLogger) *ClusterSupervisor {
+	return &ClusterSupervisor{
+		store:       store,
+		provisioner: clusterProvisioner,
+		workers:     workers,
+		logger:      logger,
+	}
+}
+
+// Do looks for work to be done on any pending clusters and attempts to schedule the required work.
+func (s *ClusterSupervisor) Do() error {
+	for {
+		more, err := s.DoOne()
+		if err != nil {
+			return errors.Wrap(err, "failed to work on cluster")
+		}
+		if !more {
+			break
+		}
+	}
+
+	return nil
+}
+
+// DoOne looks for work to be done on a single cluster, and does it.
+func (s *ClusterSupervisor) DoOne() (bool, error) {
```

Comment:
Reminder comment about what we discussed offline. Moving this to having a queue of workers waiting for their locks.